Releasedocs0 7 (#693)
microsoft/snmalloc: docs/release/0.7/README.md

# Release 0.7.0

The latest release of `snmalloc` has a few interesting features that are worth discussing in more detail.
Primarily this release focuses on improving the performance of `snmalloc` in a few key areas.
But we have also added some new features to build new security features on top of `snmalloc`.

## BatchIt

The main addition is the 0.7 release is integrating the `BatchIt` algorithm from the paper

> [BatchIt: Optimizing Message-Passing Allocators for Producer-Consumer Workloads: An Intellectual Abstract](https://dl.acm.org/doi/10.1145/3652024.3665506)
> Nathaniel Wesley Filardo, Matthew J. Parkinson

> The paper observes that for heavily producer-consumer workloads, the current `snmalloc` implementation can be suboptimal.
> The current algorithm can suffer a lot of cache-misses while it handles the messages.
> 
> BatchIt proposes to add a small producer-side cache to allow messages to the same slab of memory to be grouped.
> This means that there is much better cache locality when handling messages.

Reviewer: ```suggestion
As the title suggests, the paper considers producer-consumer workloads, wherein producer threads allocate memory for messages that get passed to consumer threads, which free their received messages.
For workloads that do significant amounts of such message passing, the `snmalloc` implementation prior to this release can be sub-optimal.
In particular, the current algorithm can suffer a lot of cache-misses while it handles returning the deallocated messages back to the producer threads' allocators.

BatchIt proposes to add a small consumer- (that is, deallocator-) side cache to allow messages to the same slab of memory to be batched together.
This results in smaller message queues within `snmalloc` and gives much better cache locality when handling messages.
```

> Nathaniel Wesley Filardo, Matthew J. Parkinson

As the title suggests, the paper considers producer-consumer workloads, wherein producer threads allocate memory for messages that get passed to consumer threads, which free their received messages.
For workloads that do significant amounts of such message passing, the `snmalloc` implementation prior to this release can be sub-optimal.
In particular, the current algorithm can suffer a lot of cache-misses while it handles returning the deallocated messages back to the producer threads' allocators.

BatchIt proposes to add a small consumer- (that is, deallocator-) side cache to allow messages to the same slab of memory to be batched together.
This results in smaller message queues within `snmalloc` and gives much better cache locality when handling messages.

We developed a micro-benchmark that simulates a producer-consumer workload with a fixed number of in-flight messages.
We then measure the time taken to process all the messages with different numbers of producer and consumer threads.
`msgpass-1` has a single producer and a single consumer, `msgpass-2` has two producers and two consumers, and so on.

> ![Graph of BatchIt performance](./snmalloc-msgpass.svg)

Reviewer: I fear a missing `git add`?

But we have also added some new features to build new security features on top of `snmalloc`.

## BatchIt

The main addition is the 0.7 release is integrating the `BatchIt` algorithm from the paper

> [BatchIt: Optimizing Message-Passing Allocators for Producer-Consumer Workloads: An Intellectual Abstract](https://dl.acm.org/doi/10.1145/3652024.3665506)
> Nathaniel Wesley Filardo, Matthew J. Parkinson

As the title suggests, the paper considers producer-consumer workloads, wherein producer threads allocate memory for messages that get passed to consumer threads, which free their received messages.
For workloads that do significant amounts of such message passing, the `snmalloc` implementation prior to this release can be sub-optimal.
In particular, the current algorithm can suffer a lot of cache-misses while it handles returning the deallocated messages back to the producer threads' allocators.

BatchIt proposes to add a small consumer- (that is, deallocator-) side cache to allow messages to the same slab of memory to be batched together.
This results in smaller message queues within `snmalloc` and gives much better cache locality when handling messages.

> We developed a micro-benchmark that simulates a producer-consumer workload with a fixed number of in-flight messages.

Reviewer: ```suggestion
We developed [a micro-benchmark](../../../src/test/perf/msgpass/msgpass.cc) that simulates a producer-consumer workload with back-pressure sending a fixed number of messages per producer.
```
